File: lib/features/billers/data/biller_catalog.dart
import 'package:freezed_annotation/freezed_annotation.dart';

import 'biller.dart';

part 'biller_catalog.freezed.dart';

@freezed
abstract class BillerCatalog with _$BillerCatalog {
  const BillerCatalog._();

  const factory BillerCatalog({
    required List<BillerCategory> categories,
    required List<Biller> billers,
  }) = _BillerCatalog;

  List<Biller> billersFor(String categoryId) =>
      billers.where((b) => b.categoryId == categoryId).toList();

  Biller? billerById(String id) =>
      billers.where((b) => b.id == id).firstOrNull;

  BillerCategory? categoryById(String id) =>
      categories.where((c) => c.id == id).firstOrNull;
}

File: src/features/billers/data/billerCatalog.ts
import type { Biller, BillerCategory } from './biller';

export interface BillerCatalog {
  readonly categories: readonly BillerCategory[];
  readonly billers: readonly Biller[];
}

export const makeCatalog = (
  categories: readonly BillerCategory[],
  billers: readonly Biller[],
): BillerCatalog => ({ categories, billers });

export const billersFor = (catalog: BillerCatalog, categoryId: string): readonly Biller[] =>
  catalog.billers.filter((b) => b.categoryId === categoryId);

export const billerById = (catalog: BillerCatalog, id: string): Biller | null =>
  catalog.billers.find((b) => b.id === id) ?? null;

export const categoryById = (catalog: BillerCatalog, id: string): BillerCategory | null =>
  catalog.categories.find((c) => c.id === id) ?? null;
